File: src/omnia/data/structure/mutable/ArrayStack.java
package omnia.data.structure.mutable;

import java.util.Iterator;
import java.util.Optional;

/** A {@link Stack} implementation backed by arrays. */
public final class ArrayStack<E> implements Stack<E> {

  private final MutableList<E> list = ArrayList.create();

  public static <E> Stack<E> create() {
    return new ArrayStack<>();
  }

  private ArrayStack() {}

  @Override
  public void push(E item) {
    list.add(item);
  }

  @Override
  public Optional<E> pop() {
    return list.isPopulated() ? Optional.of(list.removeAt(list.count())) : Optional.empty();
  }

  @Override
  public Iterator<E> iterator() {
    return list.iterator();
  }

  @Override
  public int count() {
    return list.count();
  }
}

File: src/omnia/data/structure/mutable/ArrayStack.kt
package omnia.data.structure.mutable

import java.util.Optional

/** A [Stack] implementation backed by arrays.  */
class ArrayStack<E : Any> private constructor() : Stack<E> {
  private val list: MutableList<E> = ArrayList.create()

  override fun push(item: E) {
    list.add(item)
  }

  override fun pop(): Optional<E> {
    return if (list.isPopulated) Optional.of(list.removeAt(list.count() - 1)) else Optional.empty()
  }

  override fun peek(): Optional<E> {
    return if (list.isPopulated) Optional.of(list.itemAt(list.count() - 1)) else Optional.empty()
  }

  override fun iterator(): Iterator<E> {
    return list.iterator()
  }

  override fun count(): Int {
    return list.count()
  }

  companion object {
    fun <E : Any> create(): Stack<E> {
      return ArrayStack()
    }
  }
}